### PR TITLE
gsoc26: Format Conversion Layer (Layer 2) + CLI refactor (#59, #61)

### LINKED ISSUE #59: gsoc26: Format Conversion Layer (Layer 2)
# Feature Request

**Describe the feature you'd like**
Add Layer 2 (within-class format conversion) to the Databus Python Client download pipeline. Users should be able to convert downloaded RDF and tabular files to a different serialization format in a single command using a new `--convert-format` flag.

**Why is this feature important?**
Currently the client downloads files exactly as published. If a dataset is published in Turtle but the consuming application needs N-Triples or CSV,  the user must manually convert after downloading. This feature eliminates that step and brings the Python client to feature parity with the Java Databus Client's Layer 2 implementation.

**Describe alternatives you've considered**
The Java client uses Apache Jena and Spark for this layer. For the Python 
client, rdflib is already a project dependency and natively supports all 
required formats, making it the natural choice without adding new dependencies.

**Additional context**
Equivalence classes and supported conversions:

| From Class | To Class | Formats |
|---|---|---|
| RDF Triples | RDF Triples | ntriples, turtle, rdf-xml |
| RDF Quads | RDF Quads | nquads, trig, trix, json-ld |
| Tabular | Tabular | csv, tsv |

### LINKED ISSUE #61: gsoc26: refactor cli compression
# Feature Request

**Describe the feature you'd like**  
`--convert-from` and `--convert-to` could be removed and replaced by `--compression` simply specify the target compression format, and the tool can detect the source compression from the input file extension
This is not your fault but the current CLI design. single --compression option that specifies the target compression format would be simpler and more intuitive

**Why is this feature important?**  
Workflow

### PR DESCRIPTION
**Pull request by DhanashreePetare**
# Pull Request

## Description

Implements Layer 2 (Format Conversion) for the Databus Python Client download pipeline, bringing it to feature parity with the Java client as described in Frey et al. Users can now convert between RDF serialization formats and tabular formats on-the-fly during download using the new --format flag. Also refactors the compression CLI (Issue #61) by replacing --convert-to / --convert-from with a single --compression flag.

**Related Issues**
Issue #59 (Format and Mapping Conversion Layer — Layer 2)
Issue #61 (Refactor CLI compression)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [ ] Housekeeping

## Checklist:
- [x] My code follows the [ruff code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
    - [x] `poetry run pytest` - all tests passed
    - [x] `poetry run ruff check` - no linting errors
    
## What was added:
databusclient/filehandling/format.py — Layer 2 with TripleHandler, QuadHandler, TSDHandler classes using rdflib.Graph, rdflib.Dataset, and list[list[str]] as intermediate representations respectively. Each handler exposes read(), write(), and convert().

## What was changed:

- --format flag replaces --convert-format, short aliases added (nt, ttl, rdf, xml, nq, jsonld)
- --compression replaces --convert-to / --convert-from (source auto-detected from file extension)
- Download pipeline: decompress once → convert → recompress once (was double decompress/recompress)
- Original file deleted after successful conversion
- Safe temp file handling via tempfile.NamedTemporaryFile

## Tests: 
- 9 format round trip tests in tests/test_format_round_trips.py (one per format, IR captured before conversion)

Closes #59
Closes #61

## Summary by CodeRabbit

* **New Features**
  * Added RDF format conversion support for triples and quads across multiple formats
  * Added tabular format conversion support (CSV/TSV)
  * Updated download CLI with new `--compression` and `--format` options
  * Added automatic source compression detection during downloads

* **Documentation**
  * Updated README with revised CLI documentation and usage examples

* **Tests**
  * Added comprehensive format round-trip conversion tests
  * Added test data fixtures for multiple RDF formats

### CODE REVIEW

File: databusclient/filehandling/format.py [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+            f"'{os.path.basename(input_file)}'. "
+            f"Supported extensions: {list(EXTENSION_TO_FORMAT.keys())}"
+        )
+
+    if input_format == convert_format:
+        # Input and target format are identical.
+        # Copy input to output path so the caller always receives an output file.
+        # This is important for the download pipeline which expects an output
+        # file to exist after convert_file() returns — e.g. for recompression.
+        if input_file != output_file:
```
Comment (Integer-Ctrl) [inline databusclient/filehandling/format.py:518]: The check is currently after decompressing the file. So if `input_format` and `convert_format` are identical, but the file is compressed. It is nonetheless decompressed and re-compressed which could be skipped. For example:
```
$ databusclient download https://databus.dev.dbpedia.link/fhofer/gsoc26/test-data/v1.0/test-data_count=13.nt.bz2 --format nt

Decompressing test-data_count=13.nt.bz2...
Input and target format are both 'ntriples'. Copied to output path: test-data_count=13.nt
Removed original file: test-data_count=13.nt.bz2
Recompressing test-data_count=13.nt -> test-data_count=13.nt.bz2...
```

File: databusclient/api/download.py [hunk 2]
```diff
@@ -70,11 +76,7 @@ def _should_convert_file(
         return False, None
```
Comment (Integer-Ctrl) [inline databusclient/api/download.py]: Actually, it should convert to the target compression. Currently the input file is not converted if it has no compression. For example:
```
databusclient download https://databus.dev.dbpedia.link/fhofer/gsoc26/test-data/v1.0/test-data_count=0.csv --compression bz2
SPARQL endpoint https://databus.dev.dbpedia.link/sparql
Downloading file: https://databus.dev.dbpedia.link/fhofer/gsoc26/test-data/v1.0/test-data_count=0.csv
Local directory not given, using /home/fhofer/ddrive/dev/infai/dbpedia/databus-python-client/fhofer/gsoc26/test-data/v1.0
Download file: https://databus.dev.dbpedia.link/fhofer/gsoc26/test-data/v1.0/test-data_count=0.csv
Redirects url:  https://raw.githubusercontent.com/Integer-Ctrl/gsoc-2026/refs/heads/main/test-data/sample.csv
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.08k/1.08k [00:00<00:00, 8.60MiB/s]
```
will not get compressed.

From a user-perspective I would expect to get a compressed file

Comment (Integer-Ctrl) [reply databusclient/api/download.py]: I know this is not your fault, as it was implemented that way. If it doesn't make any expense, could you correct it?

File: databusclient/api/convert.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from databusclient.filehandling.format import convert_file, get_converted_filename
+from databusclient.filehandling import mapping as _mapping
+
+from databusclient.filehandling.format import (  # noqa: F401
```
Comment (Integer-Ctrl) [inline databusclient/api/convert.py]: Any reason to keep `# noqa: F401`?
Removing the comment and the unused imports is a bit cleaner

File: pyproject.toml [hunk 4]
```diff
@@ -26,6 +26,15 @@ databusclient = "databusclient.cli:app"
 target-version = "py311"
 src = ["databusclient", "tests"]
 
+[tool.ruff.lint.per-file-ignores]
+"tests/test_format_round_trips.py" = ["F841"]
```
Comment (Integer-Ctrl) [inline pyproject.toml]: This is pretty overkill 😅 
You can remove this and in addition remove the unused variable in `tests/test_format_round_trips.py:137`

**Review by Integer-Ctrl — CHANGES_REQUESTED** [review]
One last note: Right now, the `--compression` option can only be used to create compressed files. It would be cool if there were options like `--compression none` or `--compression decompress`. **But if you'd rather work on the rest of the GSOC project first, that's fine too**. In that case, I'll just create an issue to add decompression later on. Let me know if you’d like to do this now, later, or maybe not at all. I'm fine with your preference 👍 

Overall, well done. I did discover a few edge cases, though, that need to be resolved. Once those are fixed, feel free to merge the pull request yourself if you feel confident. I trust you on that 😄 . Otherwise, let me know, and I'll test it again and merge it if everything works

**Comment by DhanashreePetare** [comment]
> One last note: Right now, the `--compression` option can only be used to create compressed files. It would be cool if there were options like `--compression none` or `--compression decompress`. **But if you'd rather work on the rest of the GSOC project first, that's fine too**. In that case, I'll just create an issue to add decompression later on. Let me know if you’d like to do this now, later, or maybe not at all. I'm fine with your preference 👍
> 
> Overall, well done. I did discover a few edge cases, though, that need to be resolved. Once those are fixed, feel free to merge the pull request yourself if you feel confident. I trust you on that 😄 . Otherwise, let me know, and I'll test it again and merge it if everything works

Thanks for the review! I'll prefer fixing all 4 comments before merging itself (will push as soon as fixed), and considering this is the 1st milestone, I think you should go ahead and merge once you're done rechecking through it, I can continue doing to for later ones. 😄
